torch/csrc/jit/codegen/cuda/root_domain_map.h

```diff
@@ -0,0 +1,331 @@
+      const TensorDomain* producer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map) const;
+
+  virtual std::ostream& print(std::ostream& os) const = 0;
+
+ protected:
+  //! Return a map between root IterDomains of a producer-consumer
+  //! pair.
+  //!
+  //! \param producer A producer TensorDomain
+  //! \param consumer A consumer TensorDomain
+  //! \param root_dims_to_map Maps only from IterDomains in this set
+  //! \param producer_to_consumer Maps from producer to consumer if true
+  virtual std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const = 0;
+};
+
+inline std::ostream& operator<<(std::ostream& os, const RootDomainMap& map) {
+  return map.print(os);
+}
+
+class TORCH_CUDA_API PairwiseRootDomainMap : public RootDomainMap {
+ public:
+  explicit PairwiseRootDomainMap(
+      const TensorView* producer,
+      const TensorView* consumer);
+
+  std::ostream& print(std::ostream& os) const override;
+
+ protected:
+  std::unordered_map<IterDomain*, IterDomain*> map(
+      const TensorDomain* producer,
+      const TensorDomain* consumer,
+      const std::unordered_set<IterDomain*>& root_dims_to_map,
+      bool producer_to_consumer) const override;
+
+  //! Creates mapping that does not consider the actual expression of
+  //! a producer-consumer. This is potentially unsafe.
+  PairwiseRootDomainMap() = default;
+
+ private:
+  const TensorView* producer_tv_ = nullptr;
+  const TensorView* consumer_tv_ = nullptr;
+  std::vector<bool> broadcast_flags_;
+};
+
+class TORCH_CUDA_API UnsafePairwiseRootDomainMap
+    : public PairwiseRootDomainMap {
+ public:
+  UnsafePairwiseRootDomainMap() : PairwiseRootDomainMap() {}
+  std::ostream& print(std::ostream& os) const override;
+};
+
+//! Represents an iteration domain of a TensorDomain. Only used for
+//! root domain mapping.
+//!
+//! Note that an IterDomain object may be reused
+//! across multiple TensorDomains, but an IterDomain in a
+//! TensorDomain may not be necessarily mappable to the same
+//! IterDomain used in a different TensorDomain. Thus, for the purpose
+//! of root domain mapping, an iteration domain needs to be identified
+//! with an IterDomain and its TensorDomain.
+class DomainKey {
+ public:
+  DomainKey() = default;
+  DomainKey(
+      const TensorDomain* td,
+      const IterDomain* id,
+      const IterDomain* concrete_id = nullptr)
+      : td_(td), id_(id), concrete_id_(concrete_id) {}
+  const TensorDomain* td() const {
+    return td_;
+  }
+  const IterDomain* id() const {
+    return id_;
+  }
+  const IterDomain* concreteId() const {
+    return concrete_id_;
+  }
+  bool operator==(const DomainKey& other) const {
+    return td() == other.td() && id() == other.id() &&
+        concreteId() == other.concreteId();
+  }
+  std::ostream& print(std::ostream& os) const;
```

Comment:
the current convention is to have debug print methods simply print to `std::cout` (so, `void print() const`)

For general "stringify" operations, I'd like to encourage everyone to adopt the `toString()` pattern I introduced with `kir::toString()`

Reply:
Sure. Should `toString` be defined outside of the class like `kir::toString()`?

Reply:
Yes, toString() should be a free (global) function, defined in the same namespace as the type it prints (so it's found through ADL - argument depended lookup: https://en.cppreference.com/w/cpp/language/adl)

NOTE: If we use this convention, there's no need to use explicit namespaces as ADL will provide the appropriate overload (so we can just say `toString(foo)` for any type we want to "stringify")